text/0000-node-api-v2.md

```diff
@@ -0,0 +1,1130 @@
+  "error": {
+    "code": -32602,
+    "message": "InvalidArgStructure \"start_height\" at position 1."
+  },
+  "id": 1,
+  "jsonrpc": "2.0"
+}
+```
+
+or method not found:
+
+```JSON
+{
+  "error": {
+    "code": -32601,
+    "message": "Method not found"
+  },
+  "id": 2,
+  "jsonrpc": "2.0"
+}
+```
+
+this list of errors is not exhaustive, for more information about the possible error objects see the [JSON RPC 2.0 specifications](https://www.jsonrpc.org/specification#error_object).
+
+
+### API Result Errors
+
+These type of errors are due to an API error during the query. This type of error is wrapped into the result. 
+The possible error objects are defined in the [`grin_api` crate](https://docs.rs/grin_api/2.0.0/grin_api/struct.Error.html)
+
+For example, trying to ban a peer that's already banned:
+
+```JSON
+{
+  "id": 2,
+  "jsonrpc": "2.0",
+  "result": {
+    "Err": {
+      "Internal": "ban peer error: PeerNotFound"
+    }
+  }
+}
+```
+
+or a block that doesn't exist:
+
+```JSON
+{
+  "id": 2,
+  "jsonrpc": "2.0",
+  "result": {
+    "Err": "NotFound"
+  }
+}
+```
+
+## Authentication
+
+Like the v1 API, the v2 API will use basic auth with the same secret. This token is usually in `grin/main/.api_secret`.
+
+## Wallet support
+
+The wallet currently uses the Grin Node API v1 so changes will be needed in the grin-wallet repository to make it compatible with the v2 Node API.
+
+## Legacy support
+
+The v1 API will remain active for a time the mode of operation for its REST API will be assumed to work as currently. This setup should allow existing wallets and apps to continue working as-is until a cutoff release for legacy mode is determined.
+
+## API only
+
+Note that this RFC does not propose making user-facing changes to the existing CLI (invoked by `grin client`) to invoke these functions. It's expected that the existing cli functionality will be modified to invoke the new API functions.
+
+## Node V1 Depreciation Timeline
+
+- v3.0.0 (January 2020): Node API v1 is marked as deprecated and will be remove in next major version.
+- v4.0.0 (July 2020): Node API v1 is removed from the code.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+The implementation of this RFC will temporarily introduce some additional code complexity as v1 and v2 node API will need to coexist until a cutoff release is determined.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+This kind of JSON-RPC API is widely used for cryptocurrencies. For instance:
+
+- [Bitcoin](https://en.bitcoin.it/wiki/API_reference_\(JSON-RPC\))
+- [Ethereum](https://github.com/ethereum/wiki/wiki/JSON-RPC)
+- [Monero](https://web.getmonero.org/resources/developer-guides/wallet-rpc.html)
+- [Tezos](https://tezos.gitlab.io/master/developer/rpc.html)
+
+# Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+This new API will simplify the deployment of new methods and will drastically simplify the work of developers by providing a clear documentation directly on docs.rs.
```

Comment:
"This API simplifies the deployment of new methods and drastically simplifies the work of developers..."